tests/unit/evaluation/test_fitness_function.py

```diff
@@ -3,56 +3,124 @@
-    fit_func = FitnessFunction(training_data)
+def test_fitness_function_has_eval_count_and_data(engine, mocker, fitness_function, training_data):
+    if engine == "Python":
+        mocker.patch.object(fitness_function, "__abstractmethods__",
+                            new_callable=set)
+        mocker.patch.object(training_data, "__abstractmethods__",
+                            new_callable=set)
+    training_data = training_data()
+    fit_func = fitness_function(training_data)
 
     assert fit_func.eval_count == 0
     assert fit_func.training_data is training_data
 
 
 @pytest.mark.parametrize("metric, expected_fit", [("mae", 1.2),
+                                                  ("mean absolute error", 1.2),
                                                   ("mse", 2.0),
-                                                  ("rmse", np.sqrt(2.0))])
-def test_vector_based_function_metrics(mocker, metric, expected_fit):
-    mocker.patch.object(VectorBasedFunction, "__abstractmethods__",
-                        new_callable=set)
-    mocker.patch.object(VectorBasedFunction, "evaluate_fitness_vector",
+                                                  ("mean squared error", 2.0),
+                                                  ("rmse", np.sqrt(2.0)),
+                                                  ("root mean squared error", np.sqrt(2.0))])
+def test_vector_based_function_metrics(engine, mocker, vector_based_function, metric, expected_fit, dummy_individual):
+    if engine == "Python":
+        mocker.patch.object(vector_based_function, "__abstractmethods__",
+                            new_callable=set)
+    mocker.patch.object(vector_based_function, "evaluate_fitness_vector",
                         return_value=[-2, -1, 0, 1, 2])
-    fit_func = VectorBasedFunction(metric=metric)
-    dummy_indv = None
+    fit_func = vector_based_function(metric=metric)
     
-    assert fit_func(dummy_indv) == pytest.approx(expected_fit)
-    fit_func.evaluate_fitness_vector.assert_called_once_with(dummy_indv)
+    assert fit_func(dummy_individual) == pytest.approx(expected_fit)
+    fit_func.evaluate_fitness_vector.assert_called_once_with(dummy_individual)
 
 
-def test_vector_based_function_invalid_metric(mocker):
-    mocker.patch.object(VectorBasedFunction, "__abstractmethods__",
-                        new_callable=set)
-    mocker.patch.object(VectorBasedFunction, "evaluate_fitness_vector",
+def test_vector_based_function_invalid_metric(engine, mocker, vector_based_function):
+    if engine == "Python":
+        mocker.patch.object(vector_based_function, "__abstractmethods__",
+                            new_callable=set)
+    mocker.patch.object(vector_based_function, "evaluate_fitness_vector",
                         return_value=[-2, -1, 0, 1, 2])
-    with pytest.raises(KeyError):
-        _ = VectorBasedFunction(metric="invalid metric")
+    with pytest.raises((KeyError, ValueError)):  # KeyError in Python, ValueError in Cpp
```

Comment:
Can we change the python error type to be consistent with c++?

Reply:
It's relatively easy to make the Python error match the C++ error and vice versa. Let me know which way you prefer (match Python or match C++), and I can fix that asap

Reply:
lets go with the ValueError

Reply:
sounds good, just changed that